### PR TITLE
ping template for Cisco IOS XR devices

### PR DESCRIPTION
**Pull request by jorlandobr**
Added a ping template for Cisco IOS XR devices. These devices begin the ping output with an extra line with a timestamp. That line would fail if tested against the cisco_ios_ping.textfsm template

### CODE REVIEW

**Comment by jorlandobr** [comment]
I don't know how to properly test the templates, I apprecciate any directions.

What I did was validating against https://textfsm.nornir.tech/

Also I did a collection of input files with ping outputs for each case (success, fail, unreachable, quench and mix) and a small program that test these inputs against a given template,

The ping output command in Cisco IOS XR devices  starts with a timestamp on the first line, so **cisco_ios_ping.textfsm** would fail to test these outputs, like this:

PS C:\python> python .\test_template.py
Error: "Could not parse line:". Rule Line: 23. Input Line: Fri Nov 17 16:28:01.206 BRA.
Error: "Could not parse line:". Rule Line: 23. Input Line: Fri Nov 17 16:29:18.710 BRA.
Error: "Could not parse line:". Rule Line: 23. Input Line: Fri Nov 17 16:29:18.710 BRA.
Error: "Could not parse line:". Rule Line: 23. Input Line: Fri Nov 17 16:29:18.710 BRA.
Error: "Could not parse line:". Rule Line: 23. Input Line: Tue Nov 14 11:03:09.626 BRA.
PS C:\python>

I modified slightly the template to use with IOS XR devices.

below are the bodies of the input files, the program and the outputs, first with the unmodified template cisco_xr_ping.textfsm and next with the modified template cisco_xr_ping_xr_mod.textfsm.

Let me know if this is enough.

**ping_xr_success.txt:**
Fri Nov 17 16:28:01.206 BRA
Type escape sequence to abort.
Sending 5, 100-byte ICMP Echos to 8.8.8.8, timeout is 2 seconds:
!!!!!
Success rate is 100 percent (5/5), round-trip min/avg/max = 7/7/7 ms

**ping_xr_fail.txt:**
Fri Nov 17 16:29:18.710 BRA
Type escape sequence to abort.
Sending 5, 100-byte ICMP Echos to 10.27.18.2, timeout is 2 seconds:
.....
Success rate is 0 percent (0/5)

**ping_xr_unreachable.txt:**
Fri Nov 17 16:29:18.710 BRA
Type escape sequence to abort.
Sending 5, 100-byte ICMP Echos to 10.227.128.2, timeout is 2 seconds:
U.U.U
Success rate is 0 percent (0/5)

**ping_xr_quench.txt:**
Fri Nov 17 16:29:18.710 BRA
Type escape sequence to abort.
Sending 5, 100-byte ICMP Echos to 10.27.18.2, timeout is 2 seconds:
QQQQQ
Success rate is 0 percent (0/5)

**ping_xr_mix.txt:**
Tue Nov 14 11:03:09.626 BRA
Type escape sequence to abort.
Sending 5, 100-byte ICMP Echos to 10.44.222.174, timeout is 2 seconds:
U.U!Q
Success rate is 20 percent (1/5), round-trip min/avg/max = 10/10/10 ms

```
import textfsm
# Load the input file to a variable
input_file = open("ping_xr_success.txt", encoding='utf-8')
raw_text_data = input_file.read()
input_file.close()

#template = open("cisco_asa_ping.textfsm")
#template = open("cisco_asa_ping_mod.textfsm")
template = open("cisco_xr_ping.textfsm")
#template = open("cisco_xr_ping_mod.textfsm")
#template = open("cisco_ios_ping.textfsm")
#template = open("cisco_ios_ping_mod.textfsm")
re_table = textfsm.TextFSM(template)

fsm_results = re_table.ParseText(raw_text_data)
print ("Printing input file ping_xr_succcess.txt" +  '\n' + str(fsm_results) + '\n' + '\n')

try:

    re_table = ''
    re_table = textfsm.TextFSM(template)
    # Load the input file to a variable
    input_file = open("ping_xr_fail.txt", encoding='utf-8')
    raw_text_data = input_file.read()
    input_file.close()
    fsm_results = re_table.ParseText(raw_text_data)
    print ("Printing input file ping_xr_fail.txt" +  '\n' + str(fsm_results) + '\n' + '\n')

except Exception as e:
    print(e)
    
try:
    
    re_table = ''
    re_table = textfsm.TextFSM(template)
    # Load the input file to a variable
    input_file = open("ping_xr_unreachable.txt", encoding='utf-8')
    raw_text_data = input_file.read()
    input_file.close()

    fsm_results = re_table.ParseText(raw_text_data)
    print ("Printing input file ping_xr_unreachable.txt" + '\n' + str(fsm_results) + '\n' + '\n')

except Exception as e:
    print(e)


try:
    
    re_table = ''
    re_table = textfsm.TextFSM(template)
    # Load the input file to a variable
    input_file = open("ping_xr_quench.txt", encoding='utf-8')
    raw_text_data = input_file.read()
    input_file.close()

    fsm_results = re_table.ParseText(raw_text_data)
    print ("Printing input file ping_xr_quench.txt" +  '\n' + str(fsm_results) + '\n' + '\n')

except Exception as e:
    print(e)

try:
    
    re_table = ''
    re_table = textfsm.TextFSM(template)
    # Load the input file to a variable
    input_file = open("ping_xr_mix.txt", encoding='utf-8')
    raw_text_data = input_file.read()
    input_file.close()

    fsm_results = re_table.ParseText(raw_text_data)
    print ("Printing input file ping_xr_mix.txt" +  '\n' + str(fsm_results) + '\n' + '\n')

except Exception as e:
    print(e)

```

**Testing with the original template:**


PS C:\python> python .\test_template.py
Printing input file ping_xr_succcess.txt
[['5', '100-byte ICMP Echos', '8.8.8.8', '2', '', ['!!!!!'], '100', '5', '7', '7', '7']]


Printing input file ping_xr_fail.txt
[['5', '100-byte ICMP Echos', '10.27.18.2', '2', '', ['.....'], '0', '0', '', '', '']]


Error: "Could not parse line:". Rule Line: 24. Input Line: U.U.U.
Error: "Could not parse line:". Rule Line: 24. Input Line: QQQQQ.
Error: "Could not parse line:". Rule Line: 24. Input Line: U.U!Q.
PS C:\python>

**Testing with the modified template:**

PS C:\python> python .\test_template.py
Printing input file ping_xr_succcess.txt
[['5', '100-byte ICMP Echos', '8.8.8.8', '2', '', ['!!!!!'], '100', '5', '7', '7', '7']]


Printing input file ping_xr_fail.txt
[['5', '100-byte ICMP Echos', '10.27.18.2', '2', '', ['.....'], '0', '0', '', '', '']]


Printing input file ping_xr_unreachable.txt
[['5', '100-byte ICMP Echos', '10.227.128.2', '2', '', ['U.U.U'], '0', '0', '', '', '']]


Printing input file ping_xr_quench.txt
[['5', '100-byte ICMP Echos', '10.27.18.2', '2', '', ['QQQQQ'], '0', '0', '', '', '']]


Printing input file ping_xr_mix.txt
[['5', '100-byte ICMP Echos', '10.44.222.174', '2', '', ['U.U!Q'], '20', '1', '10', '10', '10']]


PS C:\python>

**Comment by jorlandobr** [comment]
Added test cases (yml and raw fles) and updated index in jorlandobr-patch-3

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ on: # yamllint disable-line rule:truthy rule:comments
   push:
     branches:
       - "master"
+      - "jorlandobr-patch-2"
```
Comment (jvanderaa) [inline .github/workflows/ci.yml:10]:
```suggestion

```

Comment (jvanderaa) [reply .github/workflows/ci.yml:10]: Please also remove the patch-3 from the PR.

**Review by jvanderaa — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jorlandobr) [reply .github/workflows/ci.yml]: Removed!

Comment (jorlandobr) [reply .github/workflows/ci.yml]: > 
Closed by accident! Could yoou please reopen it and merge?

**Review by jvanderaa — APPROVED** [review]

**Comment by jvanderaa** [comment]
LGTM ,TY

**Pull request merged**